Random Walk with Restart

NicoloPosta opened the pull request:
Aggiunto il metodo per fare Random walk with restart

NicoloPosta committed:
Random Walk with Restart

Aggiunto il metodo per fare Random walk with restart

NicoloPosta committed:
Aggiornamento Random Walk

Aggiunto utilizzo di solo array e metrici sparsi

ncvescera commented on src/social_network_link_prediction/similarity_methods/global_similarity/random_walk.py:
```diff
@@ -0,0 +1,69 @@
+    random_walk_with_restart_fn = partial(random_walk_with_restart, W_normalized = W_normalized, c = c, max_iters= max_iters)
+    
+    #Run the function starting from each node of the graph
+    similarity = [random_walk_with_restart_fn(csr_matrix(e).transpose()) for e in np.identity(m)]
```
`np.identity()` ritorna una matrice identità piena. Ti serve per forza piena ?
Nel caso c'è una funzione di scipy che ritorna una matrice identità sparsa.

ncvescera replied:
Non serve per forza Full. Sparsa è meglio

ncvescera commented on src/social_network_link_prediction/similarity_methods/global_similarity/random_walk.py:
```diff
@@ -0,0 +1,69 @@
+    similarity_matrix = hstack(similarity)
+    return similarity_matrix
+
+def random_walk_with_restart(e, W_normalized, c = 0.05, max_iters = 100):    
```
`e` è una `csr_matrix`?

Sarebbe una matrice sparsa formata da una sola colonna ? 🤔

ncvescera replied:
è un `csr_array`

ncvescera commented on src/social_network_link_prediction/similarity_methods/global_similarity/random_walk.py:
```diff
@@ -0,0 +1,69 @@
+    return similarity_matrix
+
+def random_walk_with_restart(e, W_normalized, c = 0.05, max_iters = 100):    
+    old_e = e
```
Hai provato a fare una copia vera dell'oggetto ?

Ogni volta che salvi `e` (`old_e = e`) prova a fare una copia dell'oggetto e vedere se i risultati cambiano
Una roba tipo `olde_e = e.copy()`

NicoloPosta committed:
Aggiornato Random Walk

Utilizzo di array sparsi per migliorare le performance

ncvescera commented on src/social_network_link_prediction/similarity_methods/global_similarity/random_walk.py:
```diff
@@ -1,51 +1,71 @@
+def random_walk_with_restart(e, W_normalized, c = 0.05, max_iters = 100):
+    
+    # Initialize the current probability vector to the initial one and the error to 1
+    old_e = e.copy()
```
I risultati cambiano con questa operazione ?

NicoloPosta replied:
No, restano invariati

ncvescera replied:
Ok, allora puoi rimuovere quel `e.copy()` e lasciare solo `e` 🚀

NicoloPosta committed:
Migliorie gestione Moemoria

Migliorata la gestione della memoria utilizzando una matrice sparsa popolata proceduralmente anzichè l'uso di una lista.
Aggiornai i valori di ritorno delle funzioni e aggiunti commenti

ncvescera reviewed (approved):
👍🏻

CristianCosci committed:
Update random_walk.py

CristianCosci committed:
Update random_walk.py
